raft/src/main/java/org/apache/kafka/raft/FuturePurgatory.java

```diff
@@ -20,37 +20,44 @@
 
 /**
  * Simple purgatory interface which allows tracking a set of expirable futures.
+ * Each future will be associated with a comparable value which are used to determine
+ * if the future object can be completed.
+ *
+ * Note that the future objects should be organized in order so that completing awaiting
+ * futures would stop early and not traverse all the futures
  *
  * @param <T> Type completion type
  */
-public interface FuturePurgatory<T> {
+public interface FuturePurgatory<T extends Comparable<T>> {
 
     /**
      * Add a future to this purgatory for tracking.
      *
      * @param future the future tracking the expected completion. A subsequent call
-     *               to {@link #completeAll(Object)} will complete this future
+     *               to {@link #completeAll(Comparable, long)} will complete this future
      *               if it does not expire first.
+     * @param value  the comparable value of the future object that will be used to determine
+     *               if the future can be completed or not
      * @param maxWaitTimeMs the maximum time to wait for completion. If this
      *               timeout is reached, then the future will be completed exceptionally
      *               with a {@link org.apache.kafka.common.errors.TimeoutException}
      */
-    void await(CompletableFuture<T> future, long maxWaitTimeMs);
+    void await(CompletableFuture<Long> future, T value, long maxWaitTimeMs);
```

Comment:
I think we can simplify this API by letting the purgatory be responsible for creating the future:
```java
CompletableFuture<Long> future await(T value, long maxWaitTimeMs);
```
The use of the timestamp makes the API seem a little strange. Is there a strong reason not to keep `CompletableFuture<T>` so that the value gets passed through `complete`?

Reply:
For produce, I need to remember the append timestamp in order to calculate the metrics. I agree with you that this can be wrapped within the impl class but that means we need keep a `time` in the impl class and call milliseconds again 
 --- I guess it's not a big deal for complexity, just bring that up.

Reply:
I've updated the API to let the future be returned from the `await` call.